pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -1614,68 +1628,111 @@ char **NpyArr_encodeLabels(PyArrayObject *labels, JSONObjectEncoder *enc,
         ret[i] = NULL;
     }
 
-    origst = enc->start;
-    origend = enc->end;
-    origoffset = enc->offset;
-
     stride = PyArray_STRIDE(labels, 0);
     dataptr = PyArray_DATA(labels);
     getitem = (PyArray_GetItemFunc *)PyArray_DESCR(labels)->f->getitem;
     type_num = PyArray_TYPE(labels);
 
     for (i = 0; i < num; i++) {
-        if (PyTypeNum_ISDATETIME(type_num) || PyTypeNum_ISNUMBER(type_num))
-        {
-            item = (PyObject *)labels;
-            pyenc->npyType = type_num;
-            pyenc->npyValue = dataptr;
-        } else {
-            item = getitem(dataptr, labels);
-            if (!item) {
-                NpyArr_freeLabels(ret, num);
-                ret = 0;
-                break;
-            }
-        }
-
-        cLabel = JSON_EncodeObject(item, enc, labelBuffer, NPY_JSON_BUFSIZE);
-
-        if (item != (PyObject *)labels) {
-            Py_DECREF(item);
-        }
+        item = getitem(dataptr, labels);
+        if (!item) {
+	  NpyArr_freeLabels(ret, num);
+	  ret = 0;
+	  break;
+	}
+
+	// Using a date as a key we need to special case the formatting
+	if (PyTypeNum_ISDATETIME(type_num) ||
+	    PyDateTime_Check(item) || PyDate_Check(item)) {
+	  PyObject *argList = Py_BuildValue("(O)", item);
+	  if (argList == NULL) {
+	    Py_DECREF(item);
+            NpyArr_freeLabels(ret, num);
+            ret = 0;
+            break;
+	  }
 
-        if (PyErr_Occurred() || enc->errorMsg) {
+	  PyObject *ts = PyObject_CallObject(cls_timestamp, argList);
+	  Py_DECREF(argList);
+	  if (ts == NULL) {
+	    Py_DECREF(item);
+            NpyArr_freeLabels(ret, num);
+            ret = 0;
+            break;
+	  }
+
+	  if (enc->datetimeIso) {
+	    PyObject *iso = PyObject_CallMethod(ts, "isoformat", NULL);
+	    Py_DECREF(ts);
+	    if (iso == NULL) {
+	      Py_DECREF(item);
+	      NpyArr_freeLabels(ret, num);
+	      ret = 0;
+	      break;
+	    }
+
+	    cLabel = PyUnicode_AsUTF8(iso);
+	    Py_DECREF(iso);
+	    len = strlen(cLabel);
+	  } else {
+	    npy_int64 value;
+	    // TODO: refactor to not duplicate what goes on in beginTypeContext
```

Comment:
Comment says it all but this duplicates logic here:

https://github.com/pandas-dev/pandas/blob/358107330d6805829d3af8dcdd49f2e828aaa81d/pandas/_libs/src/ujson/python/objToJSON.c#L1837

I was thinking of aligning in a follow up as there's a lot of things that can be refactored but happy to do here as well